data/lua/cave_map_generator.lua

```diff
@@ -1,200 +1,184 @@
 
-local map = {}
+function callbacks.generate_map(params)
+	local map = MG.create_map(params.map_width, params.map_height, params.terrain_wall)
 
-local function loc_to_index(x,y)
-	return x + 1 + y * params.map_width
-end
-
-local function index_to_loc(index)
-	return (index -1) % params.map_width, math.floor((index -1) / params.map_width)
-end
-
-for i = 1, params.map_width * params.map_height do
-	table.insert(map, params.terrain_wall)
-end
---positions are 0-based to match the ingame locations.
-local function set_tile_index(index, val)
-	if index < 1 or index > params.map_width * params.map_height then
-		return
-	end
-	if map[index] == params.terrain_castle or map[index] == params.terrain_keep then
-		return
+	local function build_chamber(x, y, locs_set, size, jagged)
+		if locs_set:get(x,y) or not map:on_board(x, y) or size == 0 then
+			return
+		end
+		locs_set:insert(x,y)
+		for xn, yn in MG.adjacent_tiles(x, y) do
+			if random(100) <= 100 - jagged then
+				build_chamber(xn, yn, locs_set, size - 1, jagged)
+			end
+		end
 	end
-	if val == params.terrain_clear then
+
+	local function clear_tile(x, y)
+		if not map:on_board(x,y) then
+			return
+		end
+		if map:get_tile(x,y) == params.terrain_castle or map:get_tile(x,y) == params.terrain_keep then
+			return
+		end
 		local r = random(1000)
 		if r <= params.village_density then
-			map[index] = params.terrain_village
-			else
-			map[index] = params.terrain_clear
-		end
+			map:set_tile(x, y, params.terrain_village)
 		else
-		map[index] = val
-	end
-end
-
-local function set_tile(x, y, val)
-	set_tile_index(loc_to_index(x, y), val)
-end
-
-local function get_tile(x, y, val)
-	return map[loc_to_index(x,y)]
-end
-
-local function on_board(x, y)
-	return x >= 0 and y >= 0 and x < params.map_width and y < params.map_height
-end
-
-local function on_inner_board(x, y)
-	return x >= 1 and y >= 1 and x < params.map_width - 1 and y < params.map_height -  1
-end
-
-local adjacent_offset = {
-	{ {0,-1}, {1,-1}, {1,0}, {0,1}, {-1,0}, {-1,-1} },
-	{ {0,-1}, {1,0}, {1,1}, {0,1}, {-1,1}, {-1,0} }
-}
-local function adjacent_tiles(x, y)
-	local offset = adjacent_offset[2 - (x % 2)]
-	local i = 0
-	return function()
-		i = i + 1
-		if i <= 6 then
-			return offset[i][1] + x, offset[i][2] + y
-			else
-			return nil
+			map:set_tile(x, y, params.terrain_clear)
 		end
 	end
-end
 
-local function build_chamber(x, y, locs_set, size, jagged)
-	local index = loc_to_index(x, y)
-	if locs_set[index] or not on_board(x, y) or size == 0 then
-		return
-	end
-	locs_set[index] = {x, y}
-	for xn, yn in adjacent_tiles(x, y) do
-		if random(100) <= 100 - jagged then
-			build_chamber(xn, yn, locs_set, size - 1, jagged)
+	local chambers = {}
+	local chambers_by_id = {}
+	local passages = {}
+
+	for chamber in helper.child_range(params, "chamber") do
+		local chance = tonumber(chamber.chance) or 100
+		local x = chamber.x
+		local y = chamber.y
+		local id = chamber.id
+		if chance == 0 or random(100) > chance then
```

Comment:
It isn't really. I just have an aversion to needless generation of random numbers. (ie, when there can be only one possible outcome I don't see the point of bothering to generate the number.)